### PR TITLE
matrix/{cmat128,conv}: new packages

### PR DESCRIPTION
**Pull request by kortschak**
@btracey Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
When it comes time to work on cmat128 in earnest, it will be generated as described in the proposal issue #134. Currently it is just here to provide the API hook.

File: cmat128/matrix.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
+// Copyright ©2013 The gonum Authors. All rights reserved.
```
Comment (btracey) [inline cmat128/matrix.go]: 2015?

File: cmat128/matrix.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+// Untranspose returns the Matrix field.
+func (t Transpose) Untranspose() Matrix {
+	return t.Matrix
+}
+
+// Untransposer is a type that can undo an implicit transpose.
+type Untransposer interface {
+	// Note: This interface is needed to unify all of the Transpose types. In
+	// the mat64 methods, we need to test if the Matrix has been implicitly
```
Comment (btracey) [inline cmat128/matrix.go]: Mat64 -> mat128?

Comment (kortschak) [inline cmat128/matrix.go]: Thanks.

File: cmat128/matrix.go (hunk 1 above)
Comment (kortschak) [inline cmat128/matrix.go]: Copied directly from the mat64 code, so it's 2013. This also reduces diff noise when it's autogenerated.

**Comment by btracey** [comment]
I'm not convinced about the design of the types in matrix. The goal is to be able to communicate between the two packages if necessary. The goals seem to be 

- Exactly 1 copy
- Minimal number of fast paths

Am I missing an important goal?

 I would like it that the fast paths exist in some sort of copy/load function. From what I understand, the lapack functions that return complex numbers return them as 2x2 matrices, and not matrices of reals and imaginary. Even with the layout, the current design doesn't seem to encourage efficiency. It seems much simpler to me (in a global sense) to only have "translation" types. These do not implement Matrix, and so only exist to move data. This way we aren't worried about users trying to explicitly use matrix package types as input, and we can think about each package in isolation.

**Comment by kortschak** [comment]
The design goal is simplicity.

With the types here I can say something like `cm.Copy(matrix.NewComplex(m, nil))` and make use of the future cmat128.Dense.Copy method. Other approaches to interchange require that we duplicate code here.

Sure, it means that people might use these types directly in mat64 and cmat128 calls, but the worst it will do is cost them performance.

**Comment by kortschak** [comment]
*ping*

**Comment by btracey** [comment]
Having matrix import mat64 and cmat128 like this means that errors can't be moved into matrix (circular imports). If there's going to be a package imported by 64 and 128, and a second package that imports both of them, I think "matrix" should be the fundamental one imported by mat64 and mat128

**Comment by btracey** [comment]
All of the cmat128 code LGTM.

**Comment by btracey** [comment]
OK. I see the point. Maybe we could have gonum/matrix/translate that has these types, and Error can live in matrix? Something like
````
m.Copy(translate.Complex{R: a}) 
````
Translate is nice as it can also incorporate the other weird translation types, i.e. when complex numbers aren't stored two matrices but as blocks. (translate.BlockedComplex for example)

I don't think the New types are needed. matrix.Complex{R: r, I: i} seems fine. The Complex struct isn't doing anything to protect the data, as the data is returned uncopied from Real and Imag. Seems simpler and makes it clear the data isn't behind a wall. 

I don't understand what the Realer and Imager do. Who would implement them besides Complex?

**Comment by kortschak** [comment]
> Having matrix import mat64 and cmat128 like this means that errors can't be moved into matrix (circular imports).

The intention is to never import this from mat64 or cmat128.

**Comment by kortschak** [comment]
>Translate is nice as it can also incorporate the other weird translation types, i.e. when complex numbers aren't stored two matrices but as blocks. (translate.BlockedComplex for example)

Possibly trans, but it seems to be a proliferation of packages.

>I don't think the New types are needed. matrix.Complex{R: r, I: i} seems fine. The Complex struct isn't doing anything to protect the data, as the data is returned uncopied from Real and Imag. Seems simpler and makes it clear the data isn't behind a wall.

NewComplex prevents mismatched sizes from being used early with an informative panic.

>I don't understand what the Realer and Imager do. Who would implement them besides Complex? 

This prevents an import cycle.

**Comment by btracey** [comment]
Your "conv" idea is way better than translate.

How does the Error type work if it's in matrix? matrix needs to import mat64 to support Real (for example) but mat64 needs to import matrix in order to return a Condition error

**Comment by kortschak** [comment]
>Your "conv" idea is way better than translate.

I'd prefer it be here, purely to avoid an extra name space.

>How does the Error type work if it's in matrix? matrix needs to import mat64 to support Real (for example) but mat64 needs to import matrix in order to return a Condition error

I don't understand why `mat64` needs to import matrix for Real. Why does `mat64` need `Real`? The conversion routines here are for clients. If we have a function that returns a complex matrix in `mat64` (it was my understanding that we were going to return a `[]complex128` in the one case where this is an issue, but...) we can define a `Complex` type that satisfies `cmat128.Matrix` and return that. Similarly if we need to return a real matrix in `cmat128` we can define a type for it (I can see no reason for this at all though).

Is there something I'm missing?

**Comment by btracey** [comment]
It's not for Real. #275  suggests moving the Error type and common errors into this package. mat64 cannot use the error types once moved if matrix also needs to import mat64 to support Real (for instance).

**Comment by kortschak** [comment]
Ohh, yes. Sorry. Foolish. Should I make this `trans` or `conv`?

**Comment by btracey** [comment]
conv I think.

**Comment by kortschak** [comment]
PTAL

**Comment by btracey** [comment]
Sorry, I still don't understand the Realer and Imager interfaces. cmat128.Matrix and mat64.Matrix are pretty fundamentally different. In order to convert between them (say, cmat -> mat), one can just use the Real type. Are there any examples of a matrix type that would implement these interfaces that is not the Complex type? Seems like an unnecessary complexity.

LGTM otherwise

**Comment by kortschak** [comment]
Say we have a mat64 function that returns a complex matrix - this is not inconceivable. Then we can't return a cmat128.Matrix by name, so we return a mat64 interface equivalent that satisfies those interfaces.

**Comment by btracey** [comment]
Okay. LGTM.

**Pull request merged**